README.rst
.. _readme:

onETL
=====

|Repo Status| |PyPI| |PyPI License| |PyPI Python Version|
|Documentation| |Build Status| |Coverage|

.. |Repo Status| image:: https://www.repostatus.org/badges/latest/active.svg
    :target: https://github.com/MobileTeleSystems/onetl
.. |PyPI| image:: https://img.shields.io/pypi/v/onetl
    :target: https://pypi.org/project/onetl/
.. |PyPI License| image:: https://img.shields.io/pypi/l/onetl.svg
    :target: https://github.com/MobileTeleSystems/onetl/blob/develop/LICENSE.txt
.. |PyPI Python Version| image:: https://img.shields.io/pypi/pyversions/onetl.svg
    :target: https://badge.fury.io/py/onetl
.. |Documentation| image:: https://readthedocs.org/projects/onetl/badge/?version=stable
    :target: https://onetl.readthedocs.io/
.. |Build Status| image:: https://github.com/MobileTeleSystems/onetl/workflows/Tests/badge.svg
    :target: https://github.com/MobileTeleSystems/onetl/actions
.. |Coverage| image:: https://codecov.io/gh/MobileTeleSystems/onetl/branch/develop/graph/badge.svg?token=RIO8URKNZJ
    :target: https://codecov.io/gh/MobileTeleSystems/onetl

|Logo|

.. |Logo| image:: docs/static/logo_wide.svg
    :alt: onETL logo
    :target: https://github.com/MobileTeleSystems/onetl

What is onETL?
--------------

Python ETL/ELT library powered by `Apache Spark <https://spark.apache.org/>`_ & other open-source tools.

Goals
-----

* Provide unified classes to extract data from (**E**) & load data to (**L**) various stores.
* Provides `Spark DataFrame API <https://spark.apache.org/docs/3.2.0/api/python/reference/api/pyspark.sql.DataFrame.html>`_ for performing transformations (**T**) in terms of *ETL*.
* Provide direct assess to database, allowing to execute SQL queries, as well as DDL, DML, and call functions/procedures. This can be used for building up *ELT* pipelines.
* Support different `read strategies <https://onetl.readthedocs.io/en/stable/strategy/index.html>`_ for incremental and batch data fetching.
* Provide `hooks <https://onetl.readthedocs.io/en/stable/hooks/index.html>`_ & `plugins <https://onetl.readthedocs.io/en/stable/plugins.html>`_ mechanism for altering behavior of internal classes.

Non-goals
---------

* onETL is not a Spark replacement. It just provides additional functionality that Spark does not have, and simplifies UX for end users.
* onETL is not a framework, as it does not have requirements to project structure, naming, the way of running ETL/ELT processes, configuration, etc. All of that is out of scope, and should be implemented separately.
* onETL is deliberately developed without any integration with scheduling software like Apache Airflow. All integrations should be implemented as separated tools.

Requirements
------------
* **Python 3.7 - 3.11**
* PySpark 2.3.x - 3.4.x (depends on used connector)
* Java 8+ (required by Spark, see below)
* Kerberos libs & GCC (required by ``Hive`` and ``HDFS`` connectors)

Supported storages
------------------

+------------+------------+----------------------------------------------------------------------------------------------------------+
| Type       | Storage    | Powered by                                                                                               |
+============+============+==========================================================================================================+
| Database   | Clickhouse | Apache Spark `JDBC Data Source <https://spark.apache.org/docs/3.4.0/sql-data-sources-jdbc.html>`_        |
+            +------------+                                                                                                          +
|            | MSSQL      |                                                                                                          |
+            +------------+                                                                                                          +
|            | MySQL      |                                                                                                          |
+            +------------+                                                                                                          +
|            | Postgres   |                                                                                                          |
+            +------------+                                                                                                          +
|            | Oracle     |                                                                                                          |
+            +------------+                                                                                                          +
|            | Teradata   |                                                                                                          |
+            +------------+----------------------------------------------------------------------------------------------------------+
|            | Hive       | Apache Spark `Hive integration <https://spark.apache.org/docs/3.4.0/sql-data-sources-hive-tables.html>`_ |
+            +------------+----------------------------------------------------------------------------------------------------------+
|            | Greenplum  | Pivotal `Greenplum Spark connector <https://network.tanzu.vmware.com/products/vmware-tanzu-greenplum>`_  |
+            +------------+----------------------------------------------------------------------------------------------------------+
|            | MongoDB    | `MongoDB Spark connector <https://www.mongodb.com/docs/spark-connector/current>`_                        |
+------------+------------+----------------------------------------------------------------------------------------------------------+
| File       | HDFS       | `HDFS Python client <https://pypi.org/project/hdfs/>`_                                                   |
+            +------------+----------------------------------------------------------------------------------------------------------+
|            | S3         | `minio-py client <https://pypi.org/project/minio/>`_                                                     |
+            +------------+----------------------------------------------------------------------------------------------------------+
|            | SFTP       | `Paramiko library <https://pypi.org/project/paramiko/>`_                                                 |
+            +------------+----------------------------------------------------------------------------------------------------------+
|            | FTP        | `FTPUtil library <https://pypi.org/project/ftputil/>`_                                                   |
+            +------------+                                                                                                          +
|            | FTPS       |                                                                                                          |
+            +------------+----------------------------------------------------------------------------------------------------------+
|            | WebDAV     | `WebdavClient3 library <https://pypi.org/project/webdavclient3/>`_                                       |
+------------+------------+----------------------------------------------------------------------------------------------------------+


.. documentation

Documentation
-------------

See https://onetl.readthedocs.io/

.. install

How to install
---------------

.. _minimal-install:

Minimal installation
~~~~~~~~~~~~~~~~~~~~

Base ``onetl`` package contains:

* ``DBReader``, ``DBWriter`` and related classes
* ``FileDownloader``, ``FileUploader``, ``FileMover`` and related classes, like file filters & limits
* Read Strategies & HWM classes
* Plugins support

It can be installed via:

.. code:: bash

    pip install onetl

.. warning::

    This method does NOT include any connections.

    This method is recommended for use in third-party libraries which require for ``onetl`` to be installed,
    but do not use its connection classes.

.. _spark-install:

With DB connections
~~~~~~~~~~~~~~~~~~~

All DB connection classes (``Clickhouse``, ``Greenplum``, ``Hive`` and others) requires PySpark to be installed.

Firstly, you should install JDK. The exact installation instruction depends on your OS, here are some examples:

.. code:: bash

    yum install java-1.8.0-openjdk-devel  # CentOS 7 + Spark 2
    dnf install java-11-openjdk-devel  # CentOS 8 + Spark 3
    apt-get install openjdk-11-jdk  # Debian-based + Spark 3

.. _spark-compatibility-matrix:

Compatibility matrix
^^^^^^^^^^^^^^^^^^^^

+--------------------------------------------------------------+-------------+-------------+-------+
| Spark                                                        | Python      | Java        | Scala |
+==============================================================+=============+=============+=======+
| `2.3.x <https://spark.apache.org/docs/2.3.0/#downloading>`_  | 3.7 only    | 8 only      | 2.11  |
+--------------------------------------------------------------+-------------+-------------+-------+
| `2.4.x <https://spark.apache.org/docs/2.4.8/#downloading>`_  | 3.7 only    | 8 only      | 2.11  |
+--------------------------------------------------------------+-------------+-------------+-------+
| `3.2.x <https://spark.apache.org/docs/3.2.3/#downloading>`_  | 3.7 - 3.10  | 8u201 - 11  | 2.12  |
+--------------------------------------------------------------+-------------+-------------+-------+
| `3.3.x <https://spark.apache.org/docs/3.3.2/#downloading>`_  | 3.7 - 3.10  | 8u201 - 17  | 2.12  |
+--------------------------------------------------------------+-------------+-------------+-------+
| `3.4.x <https://spark.apache.org/docs/3.4.0/#downloading>`_  | 3.7 - 3.11  | 8u362 - 17  | 2.12  |
+--------------------------------------------------------------+-------------+-------------+-------+

Then you should install PySpark via passing ``spark`` to ``extras``:

.. code:: bash

    pip install onetl[spark]  # install latest PySpark

or install PySpark explicitly:

.. code:: bash

    pip install onetl pyspark==3.4.0  # install a specific PySpark version

or inject PySpark to ``sys.path`` in some other way BEFORE creating a class instance.
**Otherwise class import will fail.**


.. _files-install:

With file connections
~~~~~~~~~~~~~~~~~~~~~

All file connection classes (``FTP``,  ``SFTP``, ``HDFS`` and so on) requires specific Python clients to be installed.

Each client can be installed explicitly by passing connector name (in lowercase) to ``extras``:

.. code:: bash

    pip install onetl[ftp]  # specific connector
    pip install onetl[ftp,ftps,sftp,hdfs,s3,webdav]  # multiple connectors

To install all file connectors at once you can pass ``files`` to ``extras``:

.. code:: bash

    pip install onetl[files]

**Otherwise class import will fail.**


.. _kerberos-install:

With Kerberos support
~~~~~~~~~~~~~~~~~~~~~

Most of Hadoop instances set up with Kerberos support,
so some connections require additional setup to work properly.

* ``HDFS``
  Uses `requests-kerberos <https://pypi.org/project/requests-kerberos/>`_ and
  `GSSApi <https://pypi.org/project/gssapi/>`_ for authentication in WebHDFS.
  It also uses ``kinit`` executable to generate Kerberos ticket.

* ``Hive``
  Requires Kerberos ticket to exist before creating Spark session.

So you need to install OS packages with:

* ``krb5`` libs
* Headers for ``krb5``
* ``gcc`` or other compiler for C sources

The exact installation instruction depends on your OS, here are some examples:

.. code:: bash

    dnf install krb5-devel gcc  # CentOS, OracleLinux
    apt install libkrb5-dev gcc  # Debian-based

Also you should pass ``kerberos`` to ``extras`` to install required Python packages:

.. code:: bash

    pip install onetl[kerberos]


.. _full-install:

Full bundle
~~~~~~~~~~~

To install all connectors and dependencies, you can pass ``all`` into ``extras``:

.. code:: bash

    pip install onetl[all]

    # this is just the same as
    pip install onetl[spark,files,kerberos]

.. warning::

    This method consumes a lot of disk space, and requires for Java & Kerberos libraries to be installed into your OS.

.. quick-start

Quick start
------------

MSSQL → Hive
~~~~~~~~~~~~

Read data from MSSQL, transform & write to Hive.

.. code:: python

    # Import pyspark to initialize the SparkSession
    from pyspark.sql import SparkSession

    # import function to setup onETL logging
    from onetl.log import setup_logging

    # Import required connections
    from onetl.connection import MSSQL, Hive

    # Import onETL classes to read & write data
    from onetl.db import DBReader, DBWriter

    # change logging level to INFO, and set up default logging format and handler
    setup_logging()

    # Initiate new SparkSession
    spark = (
        SparkSession.builder.appName("spark_app_onetl_demo")
        .config("spark.jars.packages", MSSQL.package)
        .enableHiveSupport()
        .getOrCreate()
    )

    # Initiate MSSQL connection and check it
    mssql = MSSQL(
        host="mssqldb.demo.com",
        user="onetl",
        password="onetl",
        database="Telecom",
        spark=spark,
        extra={"ApplicationIntent": "ReadOnly"},
    ).check()

    # >>> INFO:|MSSQL| Connection is available.

    # Initiate reader
    reader = DBReader(
        connection=mssql,
        source="dbo.demo_table",
        columns=["on", "etl"],
        # Set some MSSQL read options:
        options=MSSQL.ReadOptions(fetchsize=10000),
    )

    # Read data to DataFrame
    df = reader.run()

    # Apply any PySpark transformations
    from pyspark.sql.functions import lit

    df_to_write = df.withColumn("engine", lit("onetl"))


    # Initiate Hive connection
    hive = Hive(cluster="rnd-dwh", spark=spark)

    # Initiate writer
    writer = DBWriter(
        connection=hive,
        target="dl_sb.demo_table",
        # Set some Hive write options:
        options=Hive.WriteOptions(mode="overwrite"),
    )

    # Write data from DataFrame to Hive
    writer.run(df_to_write)

    # Success!

SFTP → HDFS
~~~~~~~~~~~

Download files from SFTP & upload them to HDFS.

.. code:: python

    # import function to setup onETL logging
    from onetl.log import setup_logging

    # Import required connections
    from onetl.connection import SFTP, HDFS

    # Import onETL classes to download & upload files
    from onetl.file import FileDownloader, FileUploader

    # import filter & limit classes
    from onetl.file.filter import Glob, ExcludeDir
    from onetl.file.limit import MaxFilesCount

    # change logging level to INFO, and set up default logging format and handler
    setup_logging()

    # Initiate SFTP connection and check it
    sftp = SFTP(
        host="sftp.test.com",
        user="onetl",
        password="onetl",
    ).check()

    # >>> INFO:|SFTP| Connection is available.

    # Initiate downloader
    downloader = FileDownloader(
        connection=sftp,
        source_path="/remote/tests/Report",  # path on SFTP
        local_path="/local/onetl/Report",  # local fs path
        filters=[
            Glob("*.csv"),  # download only files matching the glob
            ExcludeDir(
                "/remote/tests/Report/exclude_dir/"
            ),  # exclude files from this directory
        ],
        limits=[
            MaxFilesCount(1000),  # download max 1000 files per run
        ],
        options=FileDownloader.Options(
            delete_source=True,  # delete files from SFTP after successful download
            mode="error",  # mark file as failed if it already exist in local_path
        ),
    )

    # Download files to local filesystem
    download_result = downloader.run()

    # Method run returns a DownloadResult object,
    # which contains collection of downloaded files, divided to 4 categories
    download_result

    #  DownloadResult(
    #      successful=[
    #          LocalPath('/local/onetl/Report/file_1.json'),
    #          LocalPath('/local/onetl/Report/file_2.json'),
    #      ],
    #      failed=[FailedRemoteFile('/remote/onetl/Report/file_3.json')],
    #      ignored=[RemoteFile('/remote/onetl/Report/file_4.json')],
    #      missing=[],
    #  )

    # Raise exception if there are failed files, or there were no files in the remote filesystem
    download_result.raise_if_failed() or download_result.raise_if_empty()

    # Do any kind of magic with files: rename files, remove header for csv files, ...
    renamed_files = my_rename_function(download_result.success)

    # function removed "_" from file names
    # [
    #    LocalPath('/home/onetl/Report/file1.json'),
    #    LocalPath('/home/onetl/Report/file2.json'),
    # ]

    # Initiate HDFS connection
    hdfs = HDFS(
        host="my.name.node",
        user="onetl",
        password="onetl",  # or keytab
    )

    # Initiate uploader
    uploader = FileUploader(
        connection=hdfs,
        target_path="/user/onetl/Report/",  # hdfs path
    )

    # Upload files from local fs to HDFS
    upload_result = uploader.run(renamed_files)

    # Method run returns a UploadResult object,
    # which contains collection of uploaded files, divided to 4 categories
    upload_result

    #  UploadResult(
    #      successful=[RemoteFile('/user/onetl/Report/file1.json')],
    #      failed=[FailedRemoteFile('/local/onetl/Report/file2.json')],
    #      ignored=[],
    #      missing=[],
    #  )

    # Raise exception if there are failed files, or there were no files in the local filesystem, or some input file is missing
    upload_result.raise_if_failed() or upload_result.raise_if_empty() or upload_result.raise_if_missing()


docs/changelog.rst
=========
Changelog
=========

.. toctree::
    :maxdepth: 1
    :caption: Changelog

    changelog/index


conftest.py
from __future__ import annotations

import logging
import os
import secrets
import shutil
from collections import namedtuple
from datetime import date, datetime, timedelta
from importlib import import_module
from pathlib import Path, PurePosixPath
from typing import TYPE_CHECKING
from unittest.mock import Mock

import pytest
from etl_entities import (
    Column,
    DateHWM,
    DateTimeHWM,
    FileListHWM,
    IntHWM,
    RemoteFolder,
    Table,
)
from pytest_lazyfixture import lazy_fixture

if TYPE_CHECKING:
    from pyspark.sql import SparkSession

# disable failing plugin import
os.environ["ONETL_PLUGINS_BLACKLIST"] = "failing-plugin"

from onetl.hwm.store import MemoryHWMStore
from tests.lib.common import upload_files

log = logging.getLogger(__name__)

PreparedDbInfo = namedtuple("PreparedDbInfo", ["full_name", "schema", "table"])


@pytest.fixture(scope="session")
def ftp_server():
    FTPServer = namedtuple("FTPServer", ["host", "port", "user", "password"])

    return FTPServer(
        host=os.getenv("ONETL_FTP_HOST"),
        port=os.getenv("ONETL_FTP_PORT"),
        user=os.getenv("ONETL_FTP_USER"),
        password=os.getenv("ONETL_FTP_PASSWORD"),
    )


@pytest.fixture(
    scope="function",
    params=[
        pytest.param(
            "real",
            marks=[pytest.mark.ftp, pytest.mark.file_connection, pytest.mark.connection],
        ),
    ],
)
def ftp_data(ftp_server):
    from onetl.connection import FTP

    ftp = FTP(
        host=ftp_server.host,
        port=ftp_server.port,
        user=ftp_server.user,
        password=ftp_server.password,
    )

    return ftp, PurePosixPath("/export/news_parse")


@pytest.fixture()
def ftp_connection(ftp_data):
    return ftp_data[0]


@pytest.fixture(scope="session")
def ftps_server():
    FTPSServer = namedtuple("FTPSServer", ["host", "port", "user", "password"])

    return FTPSServer(
        host=os.getenv("ONETL_FTPS_HOST"),
        port=os.getenv("ONETL_FTPS_PORT"),
        user=os.getenv("ONETL_FTPS_USER"),
        password=os.getenv("ONETL_FTPS_PASSWORD"),
    )


@pytest.fixture(
    scope="function",
    params=[
        pytest.param(
            "real",
            marks=[pytest.mark.ftps, pytest.mark.file_connection, pytest.mark.connection],
        ),
    ],
)
def ftps_data(ftps_server):
    from onetl.connection import FTPS

    ftps = FTPS(
        host=ftps_server.host,
        port=ftps_server.port,
        user=ftps_server.user,
        password=ftps_server.password,
    )

    return ftps, PurePosixPath("/export/news_parse")


@pytest.fixture()
def ftps_connection(ftps_data):
    return ftps_data[0]


@pytest.fixture(scope="session")
def s3_server():
    S3Server = namedtuple("S3Server", ["host", "port", "bucket", "access_key", "secret_key", "protocol"])

    return S3Server(
        host=os.getenv("ONETL_S3_HOST"),
        port=os.getenv("ONETL_S3_PORT"),
        bucket=os.getenv("ONETL_S3_BUCKET"),
        access_key=os.getenv("ONETL_S3_ACCESS_KEY"),
        secret_key=os.getenv("ONETL_S3_SECRET_KEY"),
        protocol=os.getenv("ONETL_S3_PROTOCOL", "http").lower(),
    )


@pytest.fixture(
    scope="function",
    params=[
        pytest.param("real", marks=[pytest.mark.s3, pytest.mark.file_connection, pytest.mark.connection]),
    ],
)
def s3_data(s3_server):
    from onetl.connection import S3

    s3 = S3(
        host=s3_server.host,
        port=s3_server.port,
        bucket=s3_server.bucket,
        access_key=s3_server.access_key,
        secret_key=s3_server.secret_key,
        protocol=s3_server.protocol,
    )

    if not s3.client.bucket_exists(s3_server.bucket):
        s3.client.make_bucket(s3_server.bucket)

    return s3, PurePosixPath("/export/news_parse")


@pytest.fixture()
def s3_connection(s3_data):
    return s3_data[0]


@pytest.fixture(scope="session")
def sftp_server():
    SFTPServer = namedtuple("SFTPServer", ["host", "port", "user", "password"])

    return SFTPServer(
        host=os.getenv("ONETL_SFTP_HOST"),
        port=os.getenv("ONETL_SFTP_PORT"),
        user=os.getenv("ONETL_SFTP_USER"),
        password=os.getenv("ONETL_SFTP_PASSWORD"),
    )


@pytest.fixture(
    scope="function",
    params=[
        pytest.param(
            "real",
            marks=[pytest.mark.sftp, pytest.mark.file_connection, pytest.mark.connection],
        ),
    ],
)
def sftp_data(sftp_server):
    from onetl.connection import SFTP

    sftp = SFTP(
        host=sftp_server.host,
        port=sftp_server.port,
        user=sftp_server.user,
        password=sftp_server.password,
    )

    return sftp, PurePosixPath("/app/news_parse")


@pytest.fixture()
def sftp_connection(sftp_data):
    return sftp_data[0]


@pytest.fixture(scope="session")
def webdav_server():
    WebDAVServer = namedtuple("WebDAVServer", ["host", "port", "user", "password", "ssl_verify", "protocol"])

    return WebDAVServer(
        host=os.getenv("ONETL_WEBDAV_HOST"),
        port=os.getenv("ONETL_WEBDAV_PORT"),
        user=os.getenv("ONETL_WEBDAV_USER"),
        password=os.getenv("ONETL_WEBDAV_PASSWORD"),
        ssl_verify=os.getenv("ONETL_WEBDAV_SSL_VERIFY", "false").lower() != "true",
        protocol=os.getenv("ONETL_WEBDAV_PROTOCOL", "http").lower(),
    )


@pytest.fixture(
    scope="session",
    params=[
        pytest.param(
            "real",
            marks=[pytest.mark.webdav, pytest.mark.file_connection, pytest.mark.connection],
        ),
    ],
)
def webdav_data(webdav_server):
    from onetl.connection import WebDAV

    webdav = WebDAV(
        host=webdav_server.host,
        port=webdav_server.port,
        user=webdav_server.user,
        password=webdav_server.password,
        ssl_verify=webdav_server.ssl_verify,
        protocol=webdav_server.protocol,
    )

    return webdav, PurePosixPath("/export/news_parse")


@pytest.fixture()
def webdav_connection(webdav_data):
    return webdav_data[0]


@pytest.fixture(
    scope="session",
    params=[
        pytest.param("real", marks=[pytest.mark.hdfs, pytest.mark.file_connection, pytest.mark.connection]),
    ],
)
def hdfs_server():
    HDFSServer = namedtuple("HDFSServer", ["host", "port"])

    return HDFSServer(
        host=os.getenv("ONETL_HDFS_HOST"),
        port=os.getenv("ONETL_HDFS_PORT"),
    )


@pytest.fixture(scope="function")
def hdfs_data(hdfs_server):
    from onetl.connection import HDFS

    hdfs = HDFS(host=hdfs_server.host, port=hdfs_server.port)
    return hdfs, PurePosixPath("/export/news_parse")


@pytest.fixture()
def hdfs_connection(hdfs_data):
    return hdfs_data[0]


@pytest.fixture(scope="function")
def resource_path(tmp_path_factory):
    original_files = Path(__file__).parent / "tests" / "resources" / "src"

    temp_dir = tmp_path_factory.mktemp("test_files") / secrets.token_hex(5)
    shutil.copytree(original_files, temp_dir)
    return temp_dir


@pytest.fixture(scope="function")
def test_files(resource_path):
    resources = resource_path / "news_parse_zp" / "2018_03_05_10_00_00"

    return [
        resources / "newsage-zp-2018_03_05_10_00_00.csv",
        resources / "newsage-zp-2018_03_05_10_10_00.csv",
    ]


@pytest.fixture(scope="function")
def upload_files_with_encoding(file_all_connections, source_path):
    local_root_filename = Path(__file__).parent / "tests" / "resources"
    remote_root_filename = source_path
    files = ["file_connection_utf.txt", "file_connection_ascii.txt"]

    for file in files:
        file_all_connections.upload_file(local_root_filename / file, remote_root_filename / file)

    return {
        "utf": remote_root_filename / "file_connection_utf.txt",
        "ascii": remote_root_filename / "file_connection_ascii.txt",
    }


@pytest.fixture(scope="session")
def warehouse_dir(tmp_path_factory):
    # https://spark.apache.org/docs/latest/sql-data-sources-hive-tables.html
    path = tmp_path_factory.mktemp("spark-warehouse")
    yield path
    shutil.rmtree(path, ignore_errors=True)


@pytest.fixture(scope="session")
def spark_metastore_dir(tmp_path_factory):
    # https://stackoverflow.com/a/44048667
    path = tmp_path_factory.mktemp("metastore_db")
    yield path
    shutil.rmtree(path, ignore_errors=True)


@pytest.fixture(scope="session")
def ivysettings_path():
    return Path(__file__).parent / "tests" / "ivysettings.xml"


@pytest.fixture(scope="session")
def spark_packages():
    import pyspark

    from onetl.connection import (
        MSSQL,
        Clickhouse,
        Greenplum,
        MongoDB,
        MySQL,
        Oracle,
        Postgres,
        Teradata,
    )

    packages = [
        Clickhouse.package,
        MSSQL.package,
        MySQL.package,
        Oracle.package,
        Postgres.package,
        Teradata.package,
    ]

    with_greenplum = os.getenv("ONETL_DB_WITH_GREENPLUM", "false").lower() == "true"

    pyspark_version = ".".join(pyspark.__version__.split(".")[:2])

    if pyspark_version == "2.3":
        if with_greenplum:
            packages.extend([Greenplum.package_spark_2_3])
        return packages

    if pyspark_version == "2.4":
        if with_greenplum:
            packages.extend([Greenplum.package_spark_2_4])
        return packages

    if pyspark_version == "3.2":
        packages.extend([MongoDB.package_spark_3_2])
        if with_greenplum:
            packages.extend([Greenplum.package_spark_3_2])
        return packages

    if pyspark_version == "3.3":
        packages.extend([MongoDB.package_spark_3_3])
        if not with_greenplum:
            return packages

        raise ValueError(f"Greenplum connector does not support Spark {pyspark.__version__}")

    if pyspark_version == "3.4":
        packages.extend([MongoDB.package_spark_3_4])
        if not with_greenplum:
            return packages

        raise ValueError(f"Greenplum connector does not support Spark {pyspark.__version__}")

    raise ValueError(f"Unsupported Spark version: {pyspark.__version__}")


@pytest.fixture(
    scope="session",
    name="spark",
    params=[
        pytest.param("real", marks=[pytest.mark.db_connection, pytest.mark.connection]),
    ],
)
def get_spark_session(warehouse_dir, spark_metastore_dir, ivysettings_path, spark_packages):
    from pyspark.sql import SparkSession

    spark = (
        SparkSession.builder.config("spark.app.name", "onetl")  # noqa: WPS221
        .config("spark.master", "local[*]")
        .config("spark.jars.packages", ",".join(spark_packages))
        .config("spark.jars.ivySettings", os.fspath(ivysettings_path))
        .config("spark.driver.memory", "1g")
        .config("spark.driver.maxResultSize", "1g")
        .config("spark.executor.cores", "1")
        .config("spark.executor.memory", "1g")
        .config("spark.executor.allowSparkContext", "true")  # Greenplum uses SparkContext on executor if master==local
        .config("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
        .config("spark.kryoserializer.buffer.max", "256m")
        .config("spark.default.parallelism", "1")
        .config("spark.driver.extraJavaOptions", f"-Dderby.system.home={os.fspath(spark_metastore_dir)}")
        .config("spark.sql.warehouse.dir", warehouse_dir)
        .enableHiveSupport()
        .getOrCreate()
    )

    yield spark
    spark.sparkContext.stop()
    spark.stop()


@pytest.fixture(
    scope="function",
    params=[
        pytest.param("mock", marks=[pytest.mark.db_connection, pytest.mark.connection]),
    ],
)
def spark_mock() -> SparkSession:
    from pyspark.sql import SparkSession

    spark = Mock(spec=SparkSession)
    spark.sparkContext = Mock()
    spark.sparkContext.appName = "abc"
    return spark


@pytest.fixture()
def processing(request, spark):
    processing_classes = {
        "clickhouse": ("tests.lib.clickhouse_processing", "ClickhouseProcessing"),
        "greenplum": ("tests.lib.greenplum_processing", "GreenplumProcessing"),
        "hive": ("tests.lib.hive_processing", "HiveProcessing"),
        "mongodb": ("tests.lib.mongodb_processing", "MongoDBProcessing"),
        "mssql": ("tests.lib.mssql_processing", "MSSQLProcessing"),
        "mysql": ("tests.lib.mysql_processing", "MySQLProcessing"),
        "oracle": ("tests.lib.oracle_processing", "OracleProcessing"),
        "postgres": ("tests.lib.postgres_processing", "PostgresProcessing"),
    }

    db_storage_name = request.function.__name__.split("_")[1]
    if db_storage_name not in processing_classes:
        raise ValueError(f"Wrong name. Please use one of: {list(processing_classes.keys())}")

    module_name, class_name = processing_classes[db_storage_name]
    module = import_module(module_name)
    db_processing = getattr(module, class_name)

    if db_storage_name == "hive":
        yield db_processing(spark)
    else:
        with db_processing() as result:
            yield result


@pytest.fixture
def get_schema_table(processing):
    schema = processing.schema
    processing.create_schema(schema=schema)

    table = f"test_{secrets.token_hex(5)}"
    full_name = f"{schema}.{table}"

    yield PreparedDbInfo(full_name=full_name, schema=schema, table=table)

    try:
        processing.drop_table(
            table=table,
            schema=schema,
        )
    except Exception:  # noqa: S110
        pass


@pytest.fixture
def prepare_schema_table(processing, get_schema_table):
    fields = {column_name: processing.get_column_type(column_name) for column_name in processing.column_names}
    _, schema, table = get_schema_table

    processing.create_table(schema=schema, table=table, fields=fields)

    return get_schema_table


@pytest.fixture
def load_table_data(prepare_schema_table, processing):
    _, schema, table = prepare_schema_table

    processing.insert_data(
        schema=schema,
        table=table,
        values=processing.create_pandas_df(),
    )

    return prepare_schema_table


@pytest.fixture(scope="function", autouse=True)
def use_memory_hwm_store(request):  # noqa: WPS325
    test_function = request.function
    entities = test_function.__name__.split("_")

    if "strategy" in entities:
        with MemoryHWMStore():
            yield None

    else:
        yield None


@pytest.fixture(
    params=[
        lazy_fixture("ftp_data"),
        lazy_fixture("ftps_data"),
        lazy_fixture("hdfs_data"),
        lazy_fixture("s3_data"),
        lazy_fixture("sftp_data"),
        lazy_fixture("webdav_data"),
    ],
)
def file_connections_data(request):
    return request.param


@pytest.fixture()
def file_all_connections(file_connections_data):
    return file_connections_data[0]


@pytest.fixture(scope="function")
def source_path(file_connections_data):
    connection, path = file_connections_data
    connection.remove_dir(path, recursive=True)
    connection.create_dir(path)
    yield path
    connection.remove_dir(path, recursive=True)


@pytest.fixture(scope="function")
def upload_test_files(file_all_connections, resource_path, source_path):
    return upload_files(resource_path, source_path, file_all_connections)


@pytest.fixture(
    params=[
        (
            IntHWM(
                source=Table(name=secrets.token_hex(5), db=secrets.token_hex(5), instance="proto://domain.com"),
                column=Column(name=secrets.token_hex(5)),
                value=10,
            ),
            5,
        ),
        (
            DateHWM(
                source=Table(name=secrets.token_hex(5), db=secrets.token_hex(5), instance="proto://domain.com"),
                column=Column(name=secrets.token_hex(5)),
                value=date(year=2023, month=8, day=15),
            ),
            timedelta(days=31),
        ),
        (
            DateTimeHWM(
                source=Table(name=secrets.token_hex(5), db=secrets.token_hex(5), instance="proto://domain.com"),
                column=Column(name=secrets.token_hex(5)),
                value=datetime(year=2023, month=8, day=15, hour=11, minute=22, second=33),
            ),
            timedelta(seconds=50),
        ),
        (
            FileListHWM(
                source=RemoteFolder(name=f"/absolute/{secrets.token_hex(5)}", instance="ftp://ftp.server:21"),
                value=["some/path", "another.file"],
            ),
            "third.file",
        ),
    ],
)
def hwm_delta(request):
    return request.param
